### DISCUSSION TITLE
Slow GPIO button detection when I2C sensors are connected

### DISCUSSION THREAD

**Discussion by daniboube** [arendst/Tasmota#18033] [Q&A]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._

I have a ESP32-C3 Luatos Board along with a LCD display hat with a joystick integrated on it. This joysticks acts like 5 GPIO buttons (left, right, up, down and center). With this setup, I've code a simple menu with some screens using Berry and LVGL, and for the button detection, I've created some rules to fire a custom command, so the user can interact with the menu. This works perfectly fine.

The problems start when I connected a TSL2591 light sensor and a AHT21 temperature and humidity sensor. Both are I2C devices. The rules that worked before have now some sort of delay, making the interaction with the screen so slow and unconfortable. Sometimes the rules are not even fired at all. 

I tried to enable option 13, change the button debounce time, but nothing worked. Is this a problem related with the chip model that I'm using? Is a I2C bus problem? Is there another way to detect gpio buttons without rules?

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [X] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [X] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [X] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [X] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [X] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [X] Device used (e.g., Sonoff Basic): ESP32-C3 Board by LuatOS
- [X] Tasmota binary firmware version number used: 12.3.1
  - [ ] Pre-compiled
  - [X] Self-compiled
- [X] Flashing tools used: esptool
- [X] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
10:07:44.389 MQT: stat/FreeDSM_9793DB0/attrs/RESULT = {"NAME":"FreeDSM CORE-ESP32C3","GPIO":[4704,1,736,704,32,33,800,768,34,35,1024,1,1,36,0,0,0,0,640,608,6210,0],"FLAG":0,"BASE":1}
10:07:44.591 MQT: stat/FreeDSM_9793DB0/attrs/RESULT = {"Module":{"0":"FreeDSM CORE-ESP32C3"}}
10:07:45.028 MQT: stat/FreeDSM_9793DB0/attrs/RESULT = {"GPIO0":{"4704":"ADC Input1"},"GPIO1":{"0":"None"},"GPIO2":{"736":"SPI CLK1"},"GPIO3":{"704":"SPI MOSI1"},"GPIO4":{"32":"Button1"},"GPIO5":{"33":"Button2"},"GPIO6":{"800":"SPI DC1"},"GPIO7":{"768":"SPI CS1"},"GPIO8":{"34":"Button3"},"GPIO9":{"35":"Button4"},"GPIO10":{"1024":"Display Rst"},"GPIO12":{"0":"None"},"GPIO13":{"36":"Button5"},"GPIO18":{"640":"I2C SDA1"},"GPIO19":{"608":"I2C SCL1"},"GPIO20":{"6210":"Option A3"},"GPIO21":{"0":"None"}}
```
- [X] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:
10:09:25.622 MQT: stat/FreeDSM_9793DB0/attrs/RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":276,"Free":235,"Rules":"ON System#Init DO Power1 1 ENDON ON System#Boot DO DeviceReady ENDON ON Button1#state DO Joystick CENTER ENDON ON Button2#state DO Joystick RIGHT ENDON ON Button3#state DO Joystick DOWN ENDON ON Button4#state DO Joystick LEFT ENDON ON Button5#state DO Joystick UP ENDON"}}
10:09:26.050 MQT: stat/FreeDSM_9793DB0/attrs/RESULT = {"Rule2":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}
10:09:26.264 MQT: stat/FreeDSM_9793DB0/attrs/RESULT = {"Rule3":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}

```
- [X] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
10:10:21.460 MQT: stat/FreeDSM_9793DB0/attrs/STATUS = {"Status":{"Module":0,"DeviceName":"FreeDSM","FriendlyName":["FreeDSM"],"Topic":"FreeDSM_9793DB0","ButtonTopic":"0","Power":1,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
10:10:21.501 MQT: stat/FreeDSM_9793DB0/attrs/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"https://ota.tasmota.com/tasmota32/release/tasmota32c3.bin","RestartReason":"Vbat power on reset","Uptime":"0T00:29:22","StartupUTC":"2023-02-23T08:40:59","Sleep":50,"CfgHolder":4617,"BootCount":42,"BCResetTime":"2023-02-17T09:38:09","SaveCount":267}}
10:10:21.542 MQT: stat/FreeDSM_9793DB0/attrs/STATUS2 = {"StatusFWR":{"Version":"12.3.1(tasmota)","BuildDateTime":"2023-02-14T13:08:11","Core":"2_0_5_3","SDK":"v4.4.3","CpuFrequency":160,"Hardware":"ESP32-C3 rev.3","CR":"454/699"}}
10:10:21.562 MQT: stat/FreeDSM_9793DB0/attrs/STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["LIA2",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["0000800B","2805C80001000600003C5A0A192800000000","00808280","00006000","00204000","00000000"]}}
10:10:21.596 MQT: stat/FreeDSM_9793DB0/attrs/STATUS4 = {"StatusMEM":{"ProgramSize":1984,"Free":895,"Heap":118,"StackLowMark":5,"PsrMax":0,"PsrFree":0,"ProgramFlashSize":4096,"FlashSize":4096,"FlashChipId":"1660CD","FlashFrequency":80,"FlashMode":"DIO","Features":["00000809","9FBAD7CF","0015A3A1","B7F7BFCF","05DA9BC0","F03E0DC7","000840F2","20200000","D43C482D","000000A1"],"Drivers":"1,2,3,4,5,7,8,9,10,11,12,13,14,16,17,20,21,24,26,27,29,34,35,38,50,52,54,59,60,62,63,86,87,88","Sensors":"1,2,3,5,6,7,8,9,10,11,12,13,14,15,17,18,19,20,21,22,26,28,31,34,37,39,40,42,43,45,51,52,55,56,57,58,59,63,64,66,67,74,85,92,95,98,127"}}
10:10:21.658 MQT: stat/FreeDSM_9793DB0/attrs/STATUS5 = {"StatusNET":{"Hostname":"FreeDSM-9793DB0-7600","IPAddress":"10.68.96.63","Gateway":"10.68.96.1","Subnetmask":"255.255.255.0","DNSServer1":"10.68.96.1","DNSServer2":"193.144.48.30","Mac":"60:55:F9:79:3D:B0","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
10:10:21.688 MQT: stat/FreeDSM_9793DB0/attrs/STATUS6 = {"StatusMQT":{"MqttHost":"193.147.36.252","MqttPort":1883,"MqttClientMask":"FreeDSM_%07X","MqttClient":"FreeDSM_9793DB0","MqttUser":"DEVICE_TELEMETRY","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
10:10:21.718 MQT: stat/FreeDSM_9793DB0/attrs/STATUS7 = {"StatusTIM":{"UTC":"2023-02-23T09:10:21","Local":"2023-02-23T10:10:21","StartDST":"2023-03-26T02:00:00","EndDST":"2023-10-29T03:00:00","Timezone":"+01:00","Sunrise":"07:44","Sunset":"18:22"}}
10:10:21.781 MQT: stat/FreeDSM_9793DB0/attrs/STATUS10 = {"StatusSNS":{"Time":"2023-02-23T10:10:21","ANALOG":{"A1":4095},"TSL2591":{"Illuminance":   66.607}}}
10:10:21.801 MQT: stat/FreeDSM_9793DB0/attrs/STATUS11 = {"StatusSTS":{"Time":"2023-02-23T10:10:21","Uptime":"0T00:29:22","UptimeSec":1762,"Heap":116,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":180,"MqttCount":1,"Berry":{"HeapUsed":26,"Objects":477},"POWER":"ON","Wifi":{"AP":1,"SSId":"LIA2","BSSId":"1C:B7:2C:35:13:C0","Channel":4,"Mode":"11n","RSSI":100,"Signal":-40,"LinkCount":1,"Downtime":"0T00:00:05"}}}

```
- [X] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:
10:46:10.012 BTN: Button1 multi-press 1
10:46:10.547 RUL: BUTTON1#STATE performs "Joystick CENTER"
10:46:10.549 SRC: Rule
10:46:10.550 CMD: Grp 0, Cmd 'JOYSTICK', Idx 1, Len 6, Pld -99, Data 'CENTER'
10:46:10.555 MQT: stat/FreeDSM_9793DB0/attrs/RESULT = {"Joystick":"Done"}
10:46:10.564 MQT: stat/FreeDSM_9793DB0/attrs/RESULT = {"Button1":{"Action":"SINGLE"}}
```

### TO REPRODUCE
_Steps to reproduce the behavior:_

With this setup, connect the I2C sensors and push any GPIO button. Yo will observe the rules being triggered in the logs with an important delay (about 4 seconds).

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

I would like to push the button and trigger the rule associated to it with minimun delay, as it works when no I2C sensors are connected.

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

Just in case, when I execute command I2CScan with the TSL2591 connected, the result is the following:
```lua
  Console output here:
10:36:52.219 MQT: stat/FreeDSM_9793DB0/attrs/RESULT = {"I2CScan":"Error 5 at 0x02"}
```

However, the TSL2591 seems to be working properly. The binary I'm using here is self-compiling, with the following user_config_override:
```c
// ----------------------------------------------
// Basic Settings

#undef  PROJECT
#define PROJECT "freedsm"

#undef  FRIENDLY_NAME
#define FRIENDLY_NAME "FreeDSM"

#ifdef ESP32
#undef FALLBACK_MODULE
#define FALLBACK_MODULE USER_MODULE
#define USER_TEMPLATE "{\"NAME\":\"FreeDSM CORE-ESP32C3\",\"GPIO\":[4704,1,736,704,32,33,800,768,34,35,1024,1,1,36,0,0,0,0,640,608,6210,0],\"FLAG\":0,\"BASE\":1}"
#endif

// ----------------------------------------------
// Telemetry Settings

#undef  TELE_PERIOD
#define TELE_PERIOD 300

#undef TELE_ON_POWER
#define TELE_ON_POWER false

// ----------------------------------------------
// Sensors Settings

#define USE_I2C

#define USE_TSL2591
#define USE_AHT2x

// ----------------------------------------------
// Display Settings

#define USE_SPI
#define USE_DISPLAY
#define USE_UNIVERSAL_DISPLAY

#define USE_BERRY 
#define USE_LVGL
#define USE_LVGL_PNG_DECODER 

// ----------------------------------------------
// Options Settings

#undef KEY_DISABLE_MULTIPRESS
#define KEY_DISABLE_MULTIPRESS true      // [SetOption1]  Disable button multipress

#undef KEY_ONLY_SINGLE_PRESS
#define KEY_ONLY_SINGLE_PRESS  true      // [SetOption13] Enable only single press to speed up button press recognition

#undef APP_DISABLE_POWERCYCLE
#define APP_DISABLE_POWERCYCLE true      // [SetOption65] Disable fast power cycle detection for device reset

#undef MQTT_BUTTONS
#define MQTT_BUTTONS           true      // [SetOption73] Detach buttons from relays and send multi-press and hold MQTT messages instead

// ----------------------------------------------
// Rules Settings

#define USE_RULES
#define USE_EXPRESSION
#define SUPPORT_IF_STATEMENT

#endif // _USER_CONFIG_OVERRIDE_H_
```

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by Jason2866** [comment]
Try without the AHT21. It is a pig on the i2c bus. The i2c implementation is faulty. AHT21 has to be used as only device on the i2c bus.

**Reply by daniboube** [reply]
I've tried with only the TSL2591 connected, and the result is slightly better, but not perfect. Still a 1-2 seconds delay on the input. Is there another aproach to detect GPIO buttons? Or maybe reducing the I2C bus accesses per second...

**Reply by Jason2866** [reply]
Does the command `i2cscan` does work without an error?
AND please update to latest v12.4.0. Buttons code is reworked.

**Reply by daniboube** [reply]
Okey, I will update and give it a try.

But no, the same error appears with both sensors...

**Reply by Jason2866** [reply]
Which Error? i2cscan?

**Reply by daniboube** [reply]
Yes, always the same message: {"I2CScan":"Error 5 at 0x02"}.

I updated the project and now it seems to work a little better, but I guess that the two sensors are too much for this board resources. And as you said, the AHT21 is a tricky sensor...

Any ideas about the I2CScan error?

**Reply by Jason2866** [reply]
No use of a AHT21 when testing. Don't use a Bread board. Use short wires. Try to be shorter than 20cm.
As long you get the i2cscan error it will not work.

**Reply by Jason2866** [reply]
The C3 Luatos board? Have one too. Works without any issues with Tasmota.
Posting a photo of your setup may helps.

**Reply by daniboube** [reply]
Of course, here you have!

![setup](https://user-images.githubusercontent.com/50242249/220927312-3c4cb9b0-2cca-433a-8fd1-6a43348c5b2f.jpg)

As you can see, the AHT21 is not connected. I have the GPIO18 as SDA (blue wire) and the GPIO19 as SCL (green wire). The TSL2591 is connected with 3.3V.  This USB-C adapter provides 5.1V. And of course, the board is located over the LCD display hat.

With the new version of Tasmota, the buttons seem to work well. There is still a 1 second delay, but I think it is fine. I don't know if having this error message when I execute the command `i2cscan` is a good thing, but the sensor is working just as I expected...

Not gonna lie, I'm a newbie in Tasmota, so maybe I'm forgetting some basic stuff...

**Reply by Jason2866** [reply]
The error with the i2cscan is a bad sign, maybe caused by the connected LC Display.
It is connected and can make issues when not configured. If not soldered remove and check if the issue is gone
You did not mentioned earlier that a Display is connected too. Every connected stuff can make trouble.

**Reply by daniboube** [reply]
If I remove the Display, the problem persists

**Reply by Jason2866** [reply]
Try first to get the display going. Or is it already working?
NVM, if the problem is without display there too. It won't help.
Can you remove the display and the config for the Display in the template?
Is now the i2cscan error gone?
Disable with `i2cdriver` every driver you don't use.

**Reply by daniboube** [reply]
I removed the Display and all the config related to it, disabled all the I2C drivers except for the 40 (the TSL2591 driver) and nothing worked. The same problem.

Maybe the chip is damaged...

**Reply by sfromis** [reply]
As long as you get that error `{"I2CScan":"Error 5 at 0x02"}` something needs fixing before trying to make the rest work.

To isolate such an issue, I'd probably make a minimal config with nothing connected, and only I2C SCL/SDA pins configured. If `I2CScan` works (no failure) with that, connect the AHT21 sensor, and if `I2CScan` already fails at this point, something bad about connections, power, pins or sensor board. Maybe try other gpio pins, if that makes a difference. Maybe pullup is not supplied on that board. Maybe a bad board. Maybe "something" else.....

If it worked with only I2C and AHT21 connected, next steps would be adding stuff slowly and in small steps, to see when it starts to fail.
